### PR TITLE
[DiagnosticVerifier] Make DiagnosticVerifier a DiagnosticConsumer

### PR DESCRIPTION
**Pull request by owenv**
Previously, `DiagnosticVerifier` hooked emission of `llvm::SMDiagnostic` to capture diagnostics for verification. This change makes `DiagnosticVerifier` a subclass of `DiagnosticConsumer` instead. This means that instead of checking expected diagnostics against the `llvm::SMDiagnostic`s emitted by `PrintedDiagnosticConsumer`, it checks them directly against the `DiagnosticInfo`s it receives from the `DiagnosticEngine`.

This is desirable for a couple reasons:
- It allows us to verify information about diagnostics that's lost when a `DiagnosticInfo` is translated to an `SMDiagnostic`. This will allow adding verifier support for educational notes in a follow-up
- It begins to reduce dependencies on `SMDiagnostic`. The current diagnostic pipeline looks like `Diagnostic -> DiagnosticInfo -> SMDiagnostic -> Text Output` when using the `PrintingDiagnosticConsumer`. Eventually I'd like to reduce that to just `Diagnostic -> DiagnosticInfo -> Text Output`. In addition to simplifying the diagnostics subsystem, this will make improving multi-byte character support in diagnostics much easier, and enable smaller QoL enhancements like https://bugs.swift.org/browse/SR-6024. (Edit: This will also allow running tests with -enable-experimental-diagnostic-formatting)

The downside of making this change now is that we lose test coverage of `PrintingDiagnosticConsumer` itself. Under the old system every test which used the verifier relied on it.

### CODE REVIEW

**Comment by owenv** [comment]
@swift-ci please test

**Comment by owenv** [comment]
Was re-reviewing this and found a few additional cleanup opportunities

**Comment by owenv** [comment]
@swift-ci please test

File: lib/Frontend/DiagnosticVerifier.cpp [hunk 1]
```diff
@@ -559,18 +468,19 @@ bool DiagnosticVerifier::verifyFile(unsigned BufferID,
 
     if (I == CapturedDiagnostics.end()) continue;
 
-    if (I->getMessage().find(expected.MessageStr) == StringRef::npos) {
+    auto columnNumber = SM.getLineAndColumn(I->Loc).second;
```
Comment (CodaFi) [inline lib/Frontend/DiagnosticVerifier.cpp]: We definitely want to use the column number from the diagnostic instance.  `llvm::SourceManager::getLineAndColumn` is going to go scanning buffers again.

Comment (owenv) [reply lib/Frontend/DiagnosticVerifier.cpp]: Good catch! We still need to call it once in `handleDiagnostic`, but that's equivalent to the scan `llvm::SMDiagnostic` would have done. I ran the test suite with both master and this branch and didn't notice any performance regressions.

**Comment by owenv** [comment]
@brentdax do you mind taking a look at this?

**Comment by brentdax** [comment]
@owenv I accidentally ended up commenting on this PR in #28964—sorry about that!

**Comment by owenv** [comment]
@brentdax No problem! I'll respond to your comments from the other PR here:

> Do we know why DiagnosticVerifier wasn't a DiagnosticConsumer before? I'm worried that this change might make the verifier not, for instance, compose with other diagnostic consumers the way it used to. Do -serialize-diagnostics-path and -emit-fixits-path in combination with -verify work the same way they used to? I'm also worried about whether other diagnostic-related features like -warnings-as-errors, -debug-diagnostic-names, or even -color-diagnostics will work the same way they used to. How do these things interact with this change?

I'm not sure why it wasn't a `DiagnosticConsumer` before. The one downside I do know of is I had to add the `suppressOutput` hack to `PrintingDiagnosticConsumer`, but I may be missing historical context somewhere.

The old verifier hooked the output of PrintingDiagnosticConsumer, and this one is independent of all the other consumers. As a result, -serialize-diagnostics-path and -emit-fixits-path continue to work because SerializedDiagnosticConsumer is unchanged. -warnings-as-errors overrides the diagnostic kind in `DiagnosticsEngine` before diagnostics reach consumers, so its behavior is also unchanged. -debug-diagnostic-names continues to work because the verifier looks for a substring instead of an exact match (and the old verifier already received the debug strings if the flag was passed). 

I hadn't thought to test the color flags, it turns out the current verifier doesn't respect them. I've updated this PR to begin taking them into account.

> Separately: Why is DiagnosticVerifier's full definition provided in this include/swift header? The other diagnostic consumers only declare a function to create an instance and hide the rest of the implementation, and the old -verify implementation also only exposed a few functions to set it up. Is there an architectural difference that forces DiagnosticVerifier to be different? If not, moving a lot of these declarations into the .cpp file, or at least into a separate header, might help us keep compile times down.

No particular reason, I looked over the code again and was able to move a decent number out of the header. `CapturedDiagnosticInfo` is still there because the vector can't have an incomplete type, and the three remaining private methods all need to be able to call `DiagnosticConsumer::getRawLoc`.

**Comment by owenv** [comment]
> Is this comment still accurate with the refactoring you've done? I don't think the final processing actually happens in this if anymore.

Yeah, this comment was in the wrong place. I've moved it

**Comment by owenv** [comment]
@swift-ci please smoke test

**Comment by owenv** [comment]
@brentdax I was able to extend the test coverage a bit to cover some of the edge cases you mentioned. Do you mind taking another look?

Also, since performance is relevant here, I ran the test suite locally with and without this change applied, and got 415s before and 430s after

**Comment by owenv** [comment]
@swift-ci please smoke test macOS

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by owenv** [comment]
@swift-ci please smoke test

**Comment by owenv** [comment]
@brentdax bump

**Comment by owenv** [comment]
@swift-ci please test

**Comment by owenv** [comment]
@swift-ci please test Windows platform

**Comment by owenv** [comment]
I'm not quite sure how to resolve these conflicts with the new dependency verifier. @CodaFi does DependencyVerifier rely on hooking the emission of `SMDiagnostic`s? I've been trying to move away from that approach in this PR because it blocks testing aspects of Swift diagnostics that can't be represented by `SMDiagnostic`, and couples the verifier to implementation details of `PrintingDiagnosticConsumer`.

**Comment by owenv** [comment]
@swift-ci please smoke test

**Comment by owenv** [comment]
@brentdax there's no rush, but do you mind taking another look at this when you have a chance? I think it's getting closer to being a viable replacement for the existing verifier

**Comment by owenv** [comment]
@swift-ci please smoke test

**Comment by owenv** [comment]
@swift-ci test

**Comment by owenv** [comment]
@swift-ci test

File: lib/Frontend/DiagnosticVerifier.cpp [hunk 2]
```diff
@@ -24,132 +24,81 @@
...
-                                   StringRef BufferName) {
+static std::vector<CapturedDiagnosticInfo>::iterator
+findDiagnostic(std::vector<CapturedDiagnosticInfo> &CapturedDiagnostics,
+               const ExpectedDiagnosticInfo &Expected, StringRef BufferName) {
   for (auto I = CapturedDiagnostics.begin(), E = CapturedDiagnostics.end();
        I != E; ++I) {
     // Verify the file and line of the diagnostic.
-    if (I->getLineNo() != (int)Expected.LineNo ||
-        I->getFilename() != BufferName)
+    if (I->Loc.isInvalid())
```
Comment (CodaFi) [inline lib/Frontend/DiagnosticVerifier.cpp]: Why this extra condition?

File: lib/Frontend/DiagnosticVerifier.cpp [hunk 3]
```diff
@@ -786,42 +771,49 @@ void DiagnosticVerifier::autoApplyFixes(unsigned BufferID,
...
+  capturedInfo.Loc = Info.Loc;
+  if (Info.Loc.isValid()) {
+    auto lineAndColumn = SM.getLineAndColumn(Info.Loc);
+    capturedInfo.Line = lineAndColumn.first;
+    capturedInfo.Column = lineAndColumn.second;
+    capturedInfo.FileName = SM.getDisplayNameForLoc(Info.Loc);
+  }
 
+  for (auto fixIt : Info.FixIts)
+    capturedInfo.FixIts.push_back(fixIt);
```
Comment (CodaFi) [inline lib/Frontend/DiagnosticVerifier.cpp]: Nit: An unbraced loop followed by a braced block can be pretty misleading. Could you please either put braces around this loop or use `std::copy`?

File: lib/FrontendTool/FrontendTool.cpp [hunk 4]
```diff
@@ -2178,6 +2175,18 @@ int swift::performFrontend(ArrayRef<const char *> Args,
     return finishDiagProcessing(1);
   }
 
+  // The verifier must be setup after CompilerInstance::setup is called above to
+  // ensure the input buffer IDs have been added.
+  std::unique_ptr<DiagnosticVerifier> verifier;
+  if (diagOpts.VerifyMode != DiagnosticOptions::NoVerify) {
```
Comment (CodaFi) [inline lib/FrontendTool/FrontendTool.cpp]: This seems like a really tricky initialization condition - maybe we should move the verifier instance onto CompilerInstance and set it up in the right place then.

File: lib/FrontendTool/FrontendTool.cpp [hunk 5]
```diff
@@ -2209,13 +2218,15 @@ int swift::performFrontend(ArrayRef<const char *> Args,
...
   if (diagOpts.VerifyMode != DiagnosticOptions::NoVerify) {
-    HadError = verifyDiagnostics(
-        Instance->getSourceMgr(),
-        Instance->getInputBufferIDs(),
-        diagOpts.VerifyMode == DiagnosticOptions::VerifyAndApplyFixes,
-        diagOpts.VerifyIgnoreUnknown);
+    // If we're in -verify mode, temporarily ignore errors. If verification
+    // failed, we'll still report an error correctly when finishDiagProcessing
+    // runs.
+    HadError = false;
```
Comment (CodaFi) [inline lib/FrontendTool/FrontendTool.cpp]: Is universally resetting the error code here a good idea? Granted, it's probably not a good idea for the verifier to have been doing this in the first place, but I imagine we've grown quite a few tests that rely on this behavior.

File: lib/Frontend/DiagnosticVerifier.cpp [hunk 6]
```diff
@@ -786,42 +771,49 @@ void DiagnosticVerifier::autoApplyFixes(unsigned BufferID,
...
-void swift::enableDiagnosticVerifier(SourceManager &SM) {
-  SM.getLLVMSourceMgr().setDiagHandler(VerifyModeDiagnosticHook,
-                                       new DiagnosticVerifier(SM));
+  {
+    llvm::raw_svector_ostream Out(capturedInfo.Message);
+    DiagnosticEngine::formatDiagnosticText(Out, Info.FormatString,
+                                           Info.FormatArgs);
+  }
+
+  CapturedDiagnostics.push_back(capturedInfo);
```
Comment (CodaFi) [inline lib/Frontend/DiagnosticVerifier.cpp]: Now that you've got this all nicely factored out, could you please compute the members of `CapturedDiagnosticInfo` separately and emplace them? This'll let you reduce the scope of the mutations being performed on the temporary, const-qualify some stuff, and make it clear which parts are contributing to the captured diagnostic and which parts are contributing to the formatting.

File: include/swift/Frontend/DiagnosticVerifier.h [hunk 7]
```diff
@@ -18,30 +18,81 @@
...
+/// This class implements support for -verify mode in the compiler.  It
+/// buffers up diagnostics produced during compilation, then checks them
+/// against expected-error markers in the source file.
+class DiagnosticVerifier : public DiagnosticConsumer {
+  SourceManager &SM;
+  std::vector<CapturedDiagnosticInfo> CapturedDiagnostics;
+  ArrayRef<unsigned> BufferIDs;
+  bool AutoApplyFixes;
+  bool IgnoreUnknown;
+  bool UseColor;
```
Comment (CodaFi) [inline include/swift/Frontend/DiagnosticVerifier.h]: Let's leave color support out of this PR to try to keep this as close as possible to a pure refactoring.

File: lib/Frontend/DiagnosticVerifier.cpp [hunk 8]
```diff
@@ -571,18 +625,18 @@ DiagnosticVerifier::verifyFile(unsigned BufferID, bool shouldAutoApplyFixes) {
...
-    if (I->getMessage().find(expected.MessageStr) == StringRef::npos) {
+    if (I->Message.find(expected.MessageStr) == StringRef::npos) {
       auto StartLoc = SMLoc::getFromPointer(expected.MessageRange.begin());
       auto EndLoc = SMLoc::getFromPointer(expected.MessageRange.end());
 
-      llvm::SMFixIt fixIt(llvm::SMRange{StartLoc, EndLoc}, I->getMessage());
+      llvm::SMFixIt fixIt(llvm::SMRange{StartLoc, EndLoc}, I->Message);
       addError(expected.MessageRange.begin(), "incorrect message found", fixIt);
-    } else if (I->getColumnNo() + 1 != (int)*expected.ColumnNo) {
+    } else if (I->Column != *expected.ColumnNo) {
```
Comment (CodaFi) [inline lib/Frontend/DiagnosticVerifier.cpp]: Were we off-by one?

File: lib/Frontend/DiagnosticVerifier.cpp [hunk 9]
```diff
@@ -647,16 +701,18 @@ DiagnosticVerifier::verifyFile(unsigned BufferID, bool shouldAutoApplyFixes) {
   bool HadUnexpectedDiag = false;
   for (unsigned i = CapturedDiagnostics.size(); i != 0; ) {
```
Comment (CodaFi) [inline lib/Frontend/DiagnosticVerifier.cpp]: Oof, I know this isn't your bug, but we should use a modern iterator idiom here.

**Review by CodaFi — CHANGES_REQUESTED** [review]
A raft of stylistic points. I really appreciate you trying to improve this long-neglected critical component.

File: lib/Frontend/DiagnosticVerifier.cpp (hunk 2 above)
Comment (owenv) [reply lib/Frontend/DiagnosticVerifier.cpp]: This was left over from a previous refactor, I removed it

File: lib/FrontendTool/FrontendTool.cpp (hunk 4 above)
Comment (owenv) [reply lib/FrontendTool/FrontendTool.cpp]: Good idea, moving it onto CompilerInstance seems to work well

File: lib/FrontendTool/FrontendTool.cpp (hunk 5 above)
Comment (owenv) [reply lib/FrontendTool/FrontendTool.cpp]: I moved this logic into `finishDiagProcessing` instead so we override the error code at the last possible moment, which I think makes this logic easier to understand.

File: lib/Frontend/DiagnosticVerifier.cpp (hunk 8 above)
Comment (owenv) [reply lib/Frontend/DiagnosticVerifier.cpp]: This is just a consequence of operating on `DiagnosticInfo` instead of `SMDiagnostic`. I don't know why their column indexing is inconsistent, but I'm hesitant to change the behavior as part of this PR

**Comment by owenv** [comment]
@CodaFi thanks for all the feedback, this is really helpful! I believe I've addressed your initial points.

**Review by CodaFi — APPROVED** [review]
Nice and tidy! Let's get this merged.

**Comment by CodaFi** [comment]
@swift-ci test

**Comment by owenv** [comment]
@swift-ci please test Windows platform

**Comment by CodaFi** [comment]
⛵️
